db/document.go

```diff
@@ -66,18 +77,52 @@ type syncData struct {
 // In its JSON form, the body's properties are at top-level while the syncData is in a special
 // "_sync" property.
 type document struct {
-	syncData
-	body Body
-	ID   string `json:"-"`
-	Cas  uint64
+	syncData        // Sync metadata
+	_body    Body   // Marshalled document body.  Unmarshalled lazily - should be accessed using Body()
+	rawBody  []byte // Raw document body, as retrieved from the bucket
+	ID       string `json:"-"` // Doc id
+	Cas      uint64 // Document cas
+}
+
+type revOnlySyncData struct {
+	casOnlySyncData
+	CurrentRev string `json:"rev"`
+}
+
+type casOnlySyncData struct {
+	Cas string `json:"cas"`
 }
 
 // Returns a new empty document.
 func newDocument(docid string) *document {
 	return &document{ID: docid, syncData: syncData{History: make(RevTree)}}
 }
 
-// Unmarshals a document from JSON data. The doc ID isn't in the data and must be given.
+// Accessors for document properties.  To support lazy unmarshalling of document contents, all access should be done through accessors
+func (doc *document) Body() Body {
+	if doc._body == nil && doc.rawBody != nil {
```

Comment:
This looks like it's not safe to call from multiple goroutines, which I guess might be true for a lot of the `document` methods that mutate.  I wonder if it's worth documenting that somewhere, or protecting with a mutex lock.

Reply:
I thought about that.  I don't think it's actually the case that document instances get references from multiple goroutines - they are typically associated with a single HTTP request.  Since the previous document.body usages didn't do any locking, I assumed the above usage was safe, and we could avoid the performance overhead of an additional mutex.

I'll add a comment to make it clear that this isn't safe for use by multiple goroutines.